### PR TITLE
Energy

### PR DESCRIPTION
**Pull request by tgiani**
Need to check that energy is implemented correctly and to improve the bootstrap function to avoid for loops (in rbm_sample.py)

### CODE REVIEW

**Comment by wilsonmr** [comment]
Hey tommy, not in today but did you make progress with the hamiltonian?

**Comment by tgiani** [comment]
Hey i may have the hamiltonian right (not sure, need to be checked) but i still didn't manage to have something working when you have more than one state together (which is needed not only for concurrent sampling but also for the implementation of the ising free energy, needed for the KL divergence). Within a couple of hours i'll push what i did on this branch so that you can check/change stuff

**Comment by wilsonmr** [comment]
yeah I can see how to do for one state, but taking inner product for multiple states might be more difficult

```

>>> import numpy as np
>>> #test with 3x3
... 
>>> N = 3
>>> a = np.zeros((N**2, N**2))
>>> b = np.arange(N**2)
>>> 
>>> #old method of reshaping and using roll
... 
>>> b_square = b.reshape((N, N))
>>> E1 = np.sum(np.roll(b_square, 1, axis=0)*b_square + 
...     np.roll(b_square, 1, axis=1)*b_square)
>>> print(E1)
318
>>> 
>>> #build hamiltonian in a
... 
>>> for i in range(N):
...     for j in range(N):
...             a[N*i + j, N*((i + 1)%N) + j] = 1
...             a[N*i + j, N*i + (j+1)%N] = 1
... 
>>> print(a)
[[ 0.  1.  0.  1.  0.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  1.  0.  0.  0.  0.]
 [ 1.  0.  0.  0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  0.  1.  0.  1.  0.]
 [ 0.  0.  0.  1.  0.  0.  0.  0.  1.]
 [ 1.  0.  0.  0.  0.  0.  0.  1.  0.]
 [ 0.  1.  0.  0.  0.  0.  0.  0.  1.]
 [ 0.  0.  1.  0.  0.  0.  1.  0.  0.]]
>>> 
>>> E2 = np.inner(b, np.inner(a, b))
>>> print(E2)
318.0
```

**Comment by tgiani** [comment]
yep.. there 's this np function which allows you to create block matrices with a chosen matrix along the diagonal . It s the function i m using in the last commit, called np.eye. I was trying to use that yesterday, donno if it could be a way

**Comment by tgiani** [comment]
Or maybe using some torch function

**Comment by wilsonmr** [comment]
I'm scanning the torch docs, can't see anything yet

**Comment by wilsonmr** [comment]
ok I think something like this could work:

H = torch.tensor(Hamiltonian)
v = stack of states(N_states * N_spins) 

```
v1 = torch.mm(v, H)
v2 = torch.cmul(v, v1)
E = torch.sum(v2, 2)
```

**Comment by wilsonmr** [comment]
so `torch.mm` does matrix mult of v (m x n) and H (n x n) where in the Z function m = 10000 and n = 64

`torch.cmul` then multiplies corresponding elements of v1 and v

then `torch.sum( v2, 2)` sums along rows, do we always give energy function a stack of states? If not this could break this unless it would accept torch.sum(v2, -1) so just sum along the final dimension which should allow both single states and stacks of states to be compatible

**Comment by wilsonmr** [comment]
also I worked out how to eliminate the loop from making hamiltonian, although I admit this probably makes negigible difference:

```
import numpy as np
from scipy.linalg import block_diag

L = 3
N = L**2
I_L = np.identity(L)
I_L_squared = np.identity(N)

a = np.roll(I_L_squared, L, axis=1) + block_diag(*([np.roll(I_L, 1, axis=1)]*L))
```
ok I haven't got minus signs but basically for a LxL lattice with N spins the hamiltonian will always be a block diagonal of L (LxL) identities which have been rolled 1 to right to hit their right neighbour plus a (NxN)  identity rolled L to the right to hit their bottom neighbour, this properly accounts for correct boundary conditions.. probs doesn't sound useful written like this but perhaps the code makes sense?

EDIT: did a short test, this is slower! Plus I think the for loop is simpler and we don't need an additional module haha

**Comment by tgiani** [comment]
Yes i m looking at it now and it looks to me that what you re saying about the torch function should work fine.. i ll try now

**Comment by wilsonmr** [comment]
@tgiani take a look at that, seems to work for the time being, need to check numbers

File: rbm_sample.py [hunk 1]
```diff
@@ -95,13 +95,14 @@ def ising_matrix():
...
+    return torch.from_numpy(a)
 
 
 
 def ising_free_energy(v, ising_matrix, nstates, beta=1.0):
-   ising_matrix = np.kron(np.eye(nstates), ising_matrix)
-   return exp(-beta*(v.dot(ising_matrix.dot(v.transpose()))))
+    v = v.double()
+    ising_matrix_mult = torch.sum(torch.mul(v,(v.mm(ising_matrix))), 1)
+    return ising_matrix_mult.mul(-beta).exp()
```
Comment (tgiani) [inline rbm_sample.py]: looks fine to me! at this point nstates can be removed

File: rbm_sample.py [hunk 2]
```diff
@@ -85,30 +85,121 @@ def ising_magnetization(field):
...
+          a[N*i+j][N*((i+1)%N)+j] -= 1
+    return torch.from_numpy(a)
+
+
+
+def ising_free_energy(v, ising_matrix, nstates, beta=1.0):
+    v = v.double()
+    ising_matrix_mult = torch.sum(torch.mul(v,(v.mm(ising_matrix))), 1)
+    return ising_matrix_mult.mul(-beta).exp()
+
```
Comment (wilsonmr) [inline rbm_sample.py]: So I think possibly we should move this to the rbm_pytorch.py once we are confident it is working and remove the exp() so it is outputting the equivalent thing to free energy, then it should be more compatible with finding Z

Comment (wilsonmr) [reply rbm_sample.py]: where to put ising_matrix is less obvious but I guess it is probably fine here for time being, since ising_free_energy just needs the hamiltonian as a tensor

File: rbm_sample.py (hunk 1 above)
Comment (wilsonmr) [reply rbm_sample.py]: done!

File: rbm_sample.py (hunk 2 above)
Comment (tgiani) [reply rbm_sample.py]: @wilsonmr it looks as if with some minor modifications it works well with the Z function. For now i've just redifined another Z just for ising but then it should be easy to define just one Z which works for both rbm and ising free energy. Yeah i think we should then move everythin in rbm_pytorch, even if all this is just ising model stuff, not strictly rbm

Comment (wilsonmr) [reply rbm_sample.py]: yeah, it's unclear to me if it should go into rbm_pytorch because that does seem to be for generic rbm stuff. I pulled your code and it seems nice, would be interesting to see if the K-L goes to zero if we use all of this stuff combined

**Comment by tgiani** [comment]
@wilsonmr  have a look at the last commit, we can restore the previous one if you don't like it. I ve added a flag in the json file. If it is tue it does the observables vs epochs stuff using the input path to access the training history, otherwise it uses that path to access the full trained machines (.last) and it does the observables vs temperature stuff. I've tested the two things separately to produce the required plots and it looks more or less ok, we need to clean up and some more tests.

**Comment by wilsonmr** [comment]
Is this ready for merge? Is there anything left to do?

**Comment by tgiani** [comment]
Still need to implement the annealing to compute Z_ising, but a part from that i thinik it can be merged to the master

File: rbm_analysis.py [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+    
+    return mag, mag_error, en, en_error, susc, susc_error, cv, cv_error
+
+
+def ising_averages_gs(mag_history, en_history, model_size, label=""):
+    """ 
+    Get observables using statistic of a single gibbs sampling.
+    In order to get the same statistic as metropolis, in the json file use:
+    - steps = 2000000
+    - save interval = 1000
```
Comment (wilsonmr) [inline rbm_analysis.py]: save interval 100

File: rbm_analysis.py [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+    - thermalization = 20000
```
Comment (wilsonmr) [inline rbm_analysis.py]: actually you have to factor in thermalization if this behaves like rbm_sample in other words if you want 20000 points then you need:
steps = 2020001
save interval = 100
thermalization = 20000
this will get you exactly 20000 points
